src/main/java/network/brightspots/rcv/CommonDataFormatReader.java

```diff
@@ -22,28 +22,91 @@
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
 import java.io.File;
+import java.io.FileInputStream;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.logging.Level;
 import javafx.util.Pair;
+import network.brightspots.rcv.CastVoteRecord.CvrParseException;
 import network.brightspots.rcv.TabulatorSession.UnrecognizedCandidatesException;
 
 class CommonDataFormatReader {
 
+  private final String STATUS_NEEDS_ADJUDICATION = "needs-adjudication";
+  private final String STATUS_NO = "no";
+  private final String BOOLEAN_TRUE = "true";
+
   private final String filePath;
   private final ContestConfig config;
-  private Map<String, Integer> unrecognizedCandidateCounts = new HashMap<>();
+  private final String contestId;
+  private final Map<String, Integer> unrecognizedCandidateCounts = new HashMap<>();
 
-  CommonDataFormatReader(String filePath, ContestConfig config) {
+  CommonDataFormatReader(String filePath, ContestConfig config, String contestId) {
     this.filePath = filePath;
     this.config = config;
+    this.contestId = contestId;
+  }
+
+  // This method will extract candidate data from a CDF file for the contestID specified in
+  // our CvrSource.
+  Map<String, String> getCandidates() throws CvrParseException {
+    Map<String, String> candidates = null;
+    if (filePath.endsWith(".xml")) {
+      candidates = getCandidatesXml();
+    } else if (filePath.endsWith(".json")) {
+      candidates = getCandidatesJson();
+    }
+    return candidates;
+  }
+
+  Map<String, String> getCandidatesXml() throws CvrParseException {
+    Map<String, String> candidates = new HashMap<>();
+    try {
+      XmlMapper xmlMapper = new XmlMapper();
+      xmlMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+      FileInputStream inputStream = new FileInputStream(new File(filePath));
+      CastVoteRecordReport cvrReport = xmlMapper.readValue(inputStream, CastVoteRecordReport.class);
+      for (Contest contest : cvrReport.Election[0].Contest) {
+        if (!contest.ObjectId.equals(this.contestId)) {
+          continue;
+        }
+        for (ContestSelection contestSelection : contest.ContestSelection) {
+          String candidateId = contestSelection.ObjectId;
+          // lookup the Candidate name
+          String candidateName = null;
+          for (Candidate candidate : cvrReport.Election[0].Candidate) {
+            if (candidate.ObjectId.equals(candidateId)) {
+              candidateName = candidate.Name;
+              // fallback to Code.Value
+              if (candidateName == null && candidate.Code != null && candidate.Code.Value != null) {
+                candidateName = candidate.Code.Value;
+              }
+              break;
+            }
+          }
+          if (candidateName == null) {
+            Logger.log(Level.WARNING, "No name found for CandidateId: %s", candidateId);
```

Comment:
is just a warning sufficient?

Reply:
Technically you can have Candidates without Name, so it's still possible to tabulate.  However, this entire code block (and the getCandidatesJson one) are going away in the next PR, so this can simply be ignored for now.  If you want a preview, check out the branch json_cdf which builds on this one.